### PR TITLE
env probe: recover Buck-target torch native libs + capture triton/fbgemm commits (schema 1.8)

### PR DESCRIPTION
**Pull request by vivekkhandelwal1**
## Summary

Two recoverability gaps surfaced while confirming `aorta env probe` against a Buck torch target (`fbcode//caffe2:torch`), where `import torch` succeeds in-process but the Python package is a link-tree and the C++ runtime is `dlopen`'d from a build-artifact dir:

- **Buck/monorepo native-lib recovery.** `libtorch_hip.so` was located only at `<torch.__file__>/../lib`, which does not exist in a Buck par layout, so `composable_kernel.pytorch_bundled`, `aotriton.*`, and `pytorch_build.binary_introspection.*` came back `null`. Added a `/proc/self/maps` fallback (`_loaded_lib_path_from_maps` / `_torch_native_lib_dir`) that recovers the real lib directory from the process's mapped libraries (torch is already imported). The original on-disk path and reason strings are preserved as the last-resort fallback, so existing "missing lib" diagnostics are unchanged when nothing is locatable. Linux-only, fail-soft.
- **Package commit capture.** Generalized `aiter`'s `+g<sha>` parse into `_extract_commit_from_version` / `_capture_python_package_commit` and added a `commit` field to the `triton` and `fbgemm` blocks. Best-effort: setuptools_scm `+g<sha>`, ROCm fork `.git<sha>`, or a `git_version`/`__commit__` module attr; `null` when no SHA (e.g. `"3.5.0+fb"`).

Schema **1.7 → 1.8**, additive nested keys only — `EnvSnapshot.from_dict` still round-trips pre-1.8 snapshots. Updates `docs/env-probe.md`.

## Test plan

- [x] `tests/instrumentation/test_environment.py` — added `TestPackageCommitExtraction` and `TestTorchNativeLibDir` (incl. a test proving the symbol dump recovers via `/proc/self/maps` for a Buck-style torch with no `<torch>/lib`); updated the 3 existing triton/fbgemm assertions + schema-version pin.
- [x] env + probe e2e + recipe suites pass (one unrelated pre-existing failure: `test_partial_false_on_clean_full_probe`, which requires `triton` installed in the venv and fails identically on `main`).
- [ ] Live validation against a real ROCm `libtorch_hip.so` on a gfx950 Buck torch target (unit-tested with a fake mapped lib; not yet exercised against a live HIP build).

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Comment by vivekkhandelwal1** [comment]
Live-validated on a gfx950 node against a real ROCm `libtorch_hip.so` (torch `2.9.1+rocm6.4`, `torch.version.hip=6.4.43484-...`, 371 MB lib):

- **maps locator** found the loaded `libtorch_hip.so` from `/proc/self/maps`.
- **Primary path** (real `<torch>/lib`): symbol counts populated (`pytorch_flash::`=4, `_efficient_attention`=18, `aotriton::`=3, `mha_fwd_aot`=1), `torch_lib_bundled={libaotriton_v2.so: true}`, aotriton `bundled_version=0.11.0` with real hash — confirms the refactor doesn't regress the normal layout.
- **Fallback path** (bogus `<torch>/lib`, lib still mapped — i.e. the Buck-target scenario): `_torch_native_lib_dir` recovered the real lib dir via `/proc/self/maps`, the `nm | c++filt` dump processed 30,482 demangled symbols, **no "not found" reasons**, and symbol counts were identical to the primary path.
- **triton commit**: this ROCm wheel reports `package_version=3.5.1` with no SHA → `commit: null` (correctly no false-positive).

So the previously-null `composable_kernel.pytorch_bundled` / `aotriton.*` / `pytorch_build.binary_introspection.*` fields now populate for a Buck torch target. Checking the live-validation box.

File: src/aorta/instrumentation/environment.py [hunk 1]
```diff
@@ -2535,6 +2578,87 @@ def _capture_pytorch_version(reasons: list[str]) -> str | None:
...
+    for attr in _PACKAGE_COMMIT_ATTRS:
+        val = getattr(mod, attr, None)
+        if isinstance(val, str) and val.strip():
+            return _extract_commit_from_version(val) or val.strip()
+    version_obj = getattr(mod, "version", None)
+    if version_obj is not None:
+        gv = getattr(version_obj, "git_version", None)
+        if isinstance(gv, str) and gv.strip():
+            return _extract_commit_from_version(gv) or gv.strip()
+    return None
```
[the remark this answers is not in the document]
Comment (vivekkhandelwal1) [reply src/aorta/instrumentation/environment.py:2921]: Good catch — fixed in d380dff. Added `_commit_from_attr_value`: a module commit attribute (`git_version`/`__commit__`/...) is now accepted only if it embeds a SHA (`+g<sha>`/`.git<sha>`) or *is* a 7-40 char hex SHA (case-insensitive, lowercased). Non-SHA values like "unknown"/"dirty"/a tag now return None, so `commit` is always a real SHA or null. Added tests covering rejection of non-SHA attrs and acceptance of a bare full SHA.

File: docs/env-probe.md [hunk 2]
```diff
@@ -201,8 +201,8 @@ unexpected failure. Callers always get back a valid, fully-shaped
...
+| `triton` | `dict` | `import triton; triton.__version__` (+ commit parse) | `package_version`, `commit` (schema 1.8). ROCm Triton fork bakes the source commit into `__version__` (e.g. `3.5.1+rocm7.2.1.gita272dfa8` -> `commit: "a272dfa8"`); fb builds versioned `+fb` carry no SHA -> `commit: null`. |
+| `fbgemm` | `dict` | optional `import fbgemm_gpu` (+ commit parse) + parse of `torch.__config__.show()` for `-DUSE_FBGEMM*` defines | `package_version`, `commit` (schema 1.8 -- best-effort git SHA from a setuptools_scm `+g<sha>` local-version segment or a `git_version`/`__commit__` module attr; `null` when fbgemm_gpu is vendored-in-torch rather than separately installed, or carries no SHA), `pytorch_use_fbgemm`, `pytorch_use_fbgemm_genai`. The two booleans capture the build-time decision baked into the PyTorch wheel even when `fbgemm_gpu` isn't a separate pip package. |
```
[the remark this answers is not in the document]
Comment (vivekkhandelwal1) [reply docs/env-probe.md:205]: Fixed in d380dff. The `schema_version` table row (and the sample probe output) now say `1.8`, and I added a `### \`1.8\` (current)` entry to the docs schema changelog (demoting 1.7).

File: src/aorta/instrumentation/environment.py [hunk 3]
```diff
@@ -2787,6 +2944,72 @@ def _probe_pytorch_bundled_ck(
...
+    except OSError:
+        return None
+    found: dict[str, Path] = {}
+    for line in text.splitlines():
+        # maps line: "addr perms offset dev inode   pathname". The
+        # pathname is optional (anonymous maps) and is the 6th field.
+        parts = line.split()
+        if len(parts) < 6:
+            continue
+        path_str = parts[5]
```
Comment (amd-vivekag) [inline src/aorta/instrumentation/environment.py:3208]: **[Low]** `line.split()` + `parts[5]` truncates any mapped path that contains a space (e.g. `/opt/my libs/libtorch_hip.so` becomes `/opt/my`), so the recovery silently misses the lib.

The pathname is always the trailing field, so split with a bounded count and take the remainder:

Fix:
```suggestion
        parts = line.split(maxsplit=5)
        if len(parts) < 6:
            continue
        path_str = parts[5]
```
(the `len(parts) < 6` check just above can be dropped since it's now folded in; adjust as needed). Real-world impact is small (torch libs rarely live under spaced paths), hence Low.

File: src/aorta/instrumentation/environment.py [hunk 4]
```diff
@@ -2535,6 +2578,120 @@ def _capture_pytorch_version(reasons: list[str]) -> str | None:
...
+
+# A bare git commit SHA: 7-40 hex chars, nothing else. Used to validate
+# the value of a module commit ATTRIBUTE (e.g. torch-style
+# ``version.git_version = "ff65f5bc..."``) before accepting it, so a
+# non-SHA attr like ``"unknown"`` / ``"dirty"`` / a tag never leaks into
+# the snapshot's ``commit`` field.
+_COMMIT_SHA_RE = re.compile(r"[0-9a-f]{7,40}", re.IGNORECASE)
+
+
+def _extract_commit_from_version(version: str | None) -> str | None:
```
Comment (amd-vivekag) [inline src/aorta/instrumentation/environment.py:2839]: **[Low]** The PR body says aiter's `+g<sha>` parse was "generalized into `_extract_commit_from_version`", but `_capture_aiter` still has its own inline `re.search(r"\+g([0-9a-f]{7,40})", version)` (~line 3432). So there are now two divergent commit parsers, and aiter's is the weaker one (it won't catch the `.git<sha>` fork form this new helper handles).

Fix: have `_capture_aiter` call `_extract_commit_from_version(version)` instead of its own regex, or drop the "generalized" wording from the PR description. Not blocking — aiter's format is `+g<sha>` so it still works today.

**Review by amd-vivekag — COMMENTED** [review]
## Review summary

**Verdict: ship** (two Low nits, nothing blocking)

Clean, well-scoped change. The two mechanisms — `/proc/self/maps` recovery of torch's native lib dir for Buck layouts, and best-effort commit extraction for `triton`/`fbgemm` — are both fail-soft, Linux-guarded, and don't touch the common wheel path (so no perf/behaviour regression when `<torch>/lib` exists). Schema bump is genuinely additive: only nested `commit` keys added, `from_dict` still round-trips pre-1.8, and `_disaster_snapshot` + the example snapshot were both updated to match. Test coverage is strong, including the key Buck-recovery case proving `nm` gets pointed at the maps-recovered path and no "not found" reason is emitted.

**Re-review of prior Copilot comments** — both Fixed (correct):
- Non-SHA module attr leaking into `commit`: resolved via `_commit_from_attr_value` (validates embed-or-is a 7-40 hex SHA), with tests rejecting `unknown`/`dirty`/tag and accepting+lowercasing a bare SHA. Verified.
- Docs still on 1.7: `schema_version` row, sample output, and changelog all bumped to 1.8. Verified.

**Findings:** 0 High · 0 Medium · 2 Low

Nice touch: preserving the original `<torch>/lib` path in the fallback reason string so the "missing lib" diagnostic still names a concrete path.

File: src/aorta/instrumentation/environment.py (hunk 3 above)
Comment (vivekkhandelwal1) [reply src/aorta/instrumentation/environment.py:3208]: Good catch — fixed in 34bff4e. `_loaded_lib_path_from_maps` now uses `line.split(maxsplit=5)`, so the pathname (always the trailing field) is kept intact as `parts[5]` even when it contains spaces (e.g. `/opt/my libs/libtorch_hip.so`). Added a regression test (`test_maps_parse_handles_spaced_pathname`) covering a spaced mapped path.

File: src/aorta/instrumentation/environment.py (hunk 4 above)
Comment (vivekkhandelwal1) [reply src/aorta/instrumentation/environment.py:2839]: Fixed in 34bff4e. `_capture_aiter` now calls `_extract_commit_from_version(version)` instead of its own inline `\+g([0-9a-f]{7,40})` regex, so there's a single commit parser and aiter also picks up the `.git<sha>` fork form.

File: src/aorta/instrumentation/environment.py [hunk 5]
```diff
@@ -2762,6 +2805,120 @@ def _capture_pytorch_version(reasons: list[str]) -> str | None:
...
+# segment, most specific first. setuptools_scm writes ``+g<sha>`` and
+# ``+<distance>.g<sha>``; the ROCm/fb forks append ``.git<sha>`` /
+# ``+git<sha>`` / ``.g<sha>`` (e.g. triton ``3.5.1+rocm7.2.1.gita272dfa8``).
+# We require a ``g``/``git`` lead-in before a 7-40 char lowercase hex run
+# so a bare local segment like ``+fb`` / ``+cpu`` / ``+rocm7.2.1`` is NOT
+# mistaken for a commit.
+_PACKAGE_COMMIT_RES: tuple[re.Pattern, ...] = (
+    re.compile(r"[+.](?:\d+\.)?git([0-9a-f]{7,40})\b"),
+    re.compile(r"[+.](?:\d+\.)?g([0-9a-f]{7,40})\b"),
+)
```
[the remark this answers is not in the document]
Comment (vivekkhandelwal1) [reply src/aorta/instrumentation/environment.py:2818]: Good catch — fixed in 6baaa0a. Both `_PACKAGE_COMMIT_RES` patterns (`+g<sha>` / `.git<sha>`) are now compiled with `re.IGNORECASE`, so a version embedding an uppercase hex SHA is captured too, aligning with `_COMMIT_SHA_RE`.

File: src/aorta/instrumentation/environment.py [hunk 6]
```diff
@@ -2762,6 +2805,120 @@ def _capture_pytorch_version(reasons: list[str]) -> str | None:
...
+    segment and the ROCm/fb fork ``.git<sha>`` / ``.g<sha>`` form.
+    Returns the lowercase-hex SHA, or ``None`` when the string carries
+    no commit (e.g. ``2.13.0a0+fb`` or ``3.5.0+cpu``). Never raises.
+    """
+    if not version:
+        return None
+    for pat in _PACKAGE_COMMIT_RES:
+        m = pat.search(version)
+        if m:
+            return m.group(1)
```
[the remark this answers is not in the document]
Comment (vivekkhandelwal1) [reply src/aorta/instrumentation/environment.py]: Fixed in 6baaa0a. Now that the patterns are case-insensitive, `_extract_commit_from_version` lowercases the captured SHA (`m.group(1).lower()`), so the return value matches the docstring ("Returns the lowercase-hex SHA") and downstream comparisons stay stable. Added a test asserting an uppercase SHA is matched and normalized to lowercase.

File: tests/instrumentation/test_environment.py [hunk 7]
```diff
@@ -3416,10 +3420,244 @@ def fake_import(name, *args, **kwargs):
...
+
+    def test_symbol_dump_recovers_via_maps_for_buck_torch(
+        self, isolated_env, tmp_path, monkeypatch
+    ):
+        """The key Buck recovery: no <torch>/lib/libtorch_hip.so on disk,
+        but the lib is mapped into the process -> the symbol dump finds it
+        via /proc/self/maps instead of recording a 'not found' reason.
+        """
+        import builtins
+        import types
```
[the remark this answers is not in the document]
Comment (vivekkhandelwal1) [reply tests/instrumentation/test_environment.py]: Fixed in ff8060d. Removed the unused `import builtins` from `test_symbol_dump_recovers_via_maps_for_buck_torch`; the test only uses `types`, so this clears the pyflakes noise.

**Review by amd-vivekag — APPROVED** [review]
No further review comments - I verified the prior human and Copilot findings are fixed; approving.
